src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -0,0 +1,1468 @@
+  }
+}
+
+template <typename I>
+void Replayer<I>::handle_regular_snapshot_complete(
+    int r, const std::string &group_snap_id, Context *on_finish) {
+  dout(10) << group_snap_id << ", r=" << r << dendl;
+  on_finish->complete(r);
+}
+
+template <typename I>
+void Replayer<I>::shut_down(Context* on_finish) {
+  dout(10) << dendl;
+
+  {
+    std::unique_lock locker{m_lock};
+    m_stop_requested = false;
+    ceph_assert(m_on_shutdown == nullptr);
+    std::swap(m_on_shutdown, on_finish);
+
+    auto state = STATE_COMPLETE;
+    std::swap(m_state, state);
+  }
+
+  cancel_load_group_snapshots();
+
+  if (!m_in_flight_op_tracker.empty()) {
+    m_in_flight_op_tracker.wait_for_ops(new LambdaContext([this](int) {
+        finish_shut_down();
+      }));
+    return;
+  }
+
+  finish_shut_down();
+  return;
+}
+
+template <typename I>
+void Replayer<I>::finish_shut_down() {
+  dout(10) << dendl;
+
+  Context *on_finish = nullptr;
+
+  {
+    std::unique_lock locker{m_lock};
+    ceph_assert(m_on_shutdown != nullptr);
+    std::swap(m_on_shutdown, on_finish);
+  }
+  if (on_finish) {
+    on_finish->complete(0);
+  }
+}
+
+
+template <typename I>
+bool Replayer<I>::get_replay_status(std::string* description) {
+  dout(10) << dendl;
+
+  std::unique_lock locker{m_lock};
+  if (m_state != STATE_REPLAYING) {
+    derr << "replay not running" << dendl;
+    return false;
+  }
+
+  json_spirit::mObject root_obj;
+  root_obj["last_snapshot_complete_seconds"] =
+    m_last_snapshot_complete_seconds;
+  root_obj["last_snapshot_bytes"] = m_last_snapshot_bytes;
+
+  auto remote_gp_snap_ptr = get_latest_mirror_group_snapshot(
+    m_remote_group_snaps);
+  if (remote_gp_snap_ptr != nullptr) {
+    utime_t timestamp;
+    int r = get_group_snapshot_timestamp(m_remote_io_ctx, *remote_gp_snap_ptr,
+                                         &timestamp);
+    if (r < 0) {
+      derr << "error getting timestamp of remote group snapshot ID: "
+           << remote_gp_snap_ptr->id << ", r=" << cpp_strerror(r) << dendl;
+      return false;
+    } else {
+      root_obj["remote_snapshot_timestamp"] = timestamp.sec();
+    }
+  } else {
+    return false;
+  }
+
+  auto latest_local_gp_snap_ptr = get_latest_mirror_group_snapshot(
+    m_local_group_snaps);
+  if (latest_local_gp_snap_ptr != nullptr) {
+    // find remote group snap with ID matching that of the latest local
+    // group snap
+    remote_gp_snap_ptr = nullptr;
+    for (const auto& remote_group_snap: m_remote_group_snaps) {
+      if (remote_group_snap.id == latest_local_gp_snap_ptr->id) {
+	remote_gp_snap_ptr = &remote_group_snap;
+	break;
+      }
+    }
+    if (remote_gp_snap_ptr != nullptr) {
+      utime_t timestamp;
+      int r = get_group_snapshot_timestamp(m_remote_io_ctx,
+                                           *remote_gp_snap_ptr,
+                                           &timestamp);
+      if (r < 0) {
+	derr << "error getting timestamp of remote group snapshot ID: "
```

Comment:
Make it so that these errors can be distinguished:
```suggestion
	derr << "error getting timestamp of matching remote group snapshot ID: "
```